### PR TITLE
feat: Changes for PKI support - #2694

### LINKED ISSUE #2694: Send a PKI from the client to the server
Need a way to send a PKI from the client to the server when it is requested by the server on login (two way SSL).

On connecting to our application, the server requests a PKI / personal certificate from the client.  This PKI is unique to the user; it is installed in the browser so that they can authenticate to websites with it.

The browser prompts the user to “Select a certificate to authenticate yourself”, where the user can select OK to send it.

This prompt/dialog isn’t appearing in Test Runner.  Instead, our custom error message indicating that the user did not send a PKI is displayed.

I attempted to use a group policy to force all instances of Chrome to auto select a certificate, but the PKI still isn’t sent from the client to the server. The group policy does work for all other instances of Chrome that aren’t created by Cypress.

### PR DESCRIPTION
**Pull request by GCHQDeveloper911**
- Closes #2694
### User facing changelog
Adds support for accessing websites which require authentication using PKI certificates
### Additional details
Enterprises are moving towards having users identified by managed certificates and having Cypress natively support presenting such identities will enable wider usage without additional proxies etc.
The aim of these changes is to add PKI support into a common point within the Cypress network processing so that is available to all higher-level interfaces and doesn't require changes to existing APIs / tests. Usage (or not) of certificates is configurable via cypress.json.
A new optional section is added to cypress.json that defines the PKI cert(s) to use for host URLs. These can be matched explicitly or a wildcard definition can be used; the '*' will match any character in a regex manner.
Both PEM (.crt, .pem) and PFX (.pfx) files are supported with (if required) pass-phrases supplied in associated text files.  These can be specified relative to the Cypress test project root or via an absolute path.
```json
{
  "clientCertificates": [
    {
      "url": "https://a.host.com",
      "ca": [
          "certs/ca.pem"
      ],
      "certs": [
        {
          "cert": "certs/cert.pem",
          "key": "certs/private.key",
          "passphrase": "certs/pem-passphrase.txt"
        }
      ]
    },
    {
      "url": "https://b.host.com/a_base_route/*",
      "ca": [],
      "certs": [
        {
          "pfx": "/home/tester/certs/cert.pfx",
          "passphrase": "/home/tester/certs/pfx-passphrase.txt"
        }
      ]
    },
    {
      "url": "https://a.host.*.com/",
      "certs": [
        {
          "pfx": "certs/cert.pfx",
          "passphrase": "certs/pfx-passphrase.txt"
        }
      ]
    }
  ]
}
```
#### Additions and changes
* /cli/schema/cypress.schema.json - Added new config items
* /packages/server/lib/util/validation.js - Added validation for new config values
* /packages/server/lib/pki.ts - New package that loads/parses PKI certs based on config
* /packages/server/lib/config_options.ts - Add new config settings
* /packages/server/lib/config_options.ts - Add new config options
* /packages/server/lib/util/validation.js - Add new config validation
* /packages/server/lib/test/unit/config_spec.ts - New unit tests
* /packages/network/lib/pki.ts - New package for matching URLs to certificates
* /packages/network/lib/test/unit/pki_spec.ts - New unit tests
* /packages/network/lib/agent.ts - Updated addRequest() to add PKI certs to all requests where the URL matches one defined in config
#### Testing
The code has been initially tested with:
* React applications retrieving data from backend services that require client certificates
* Via Chrome, FireFox, Edge and Electron browsers
* Using cy.request() from within tests to directly communicate with backend services that require client certificates
* We have created a sample project to demonstrate its functionality - https://github.com/GCHQDeveloper911/cypress-pki-sample-project
### How has the user experience changed?
User experience doesn't change, except for PKI-protected sites are now accessible when certificates are configured. These load and behave as any other content in Cypress would.
### PR Tasks
- [x] Have tests been added/updated?
- [ ] Has the original issue or this PR been tagged with a release in ZenHub? 
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [x] Have new configuration options been added to the [`cypress.schema.json`](https://github.com/cypress-io/cypress/blob/develop/cli/schema/cypress.schema.json)?

### CODE REVIEW

**Comment by jennifer-shehane** [comment]
Thanks @GCHQDeveloper911, we're going to try to pull in an engineer to take a look at this next sprint since we're prepping for 7.0.0 release now. Please be patient if we don't review right away. 🙏  It looks like there's a few lint issues atm.

**Comment by GCHQDeveloper911** [comment]
> Thanks @GCHQDeveloper911, we're going to try to pull in an engineer to take a look at this next sprint since we're prepping for 7.0.0 release now. Please be patient if we don't review right away. 🙏 It looks like there's a few lint issues atm.

Pleasure - no worries on the delay... We'll keep checking the CI results and fix any issues meantime. Many thanks!

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane - Can you advise here, it seems the final 9 checks are stuck? In the previous run (commit b7a5d62) - I think these all ran successfully?

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 I dunno, I see this happen sometimes. I'll pull in develop and rerun.

**Comment by GCHQDeveloper911** [comment]
Thanks for that. Looking at that e2e-tests-firefox fail:

Error: Expected video to have seekable ending frame, but it did not. The video may be corrupted.
    at /root/cypress/packages/server/test/e2e/6_video_compression_spec.js:67:21
    at /root/cypress/packages/server/test/e2e/6_video_compression_spec.js:66:13
    
I think it's safe to say that's not our changes?

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane - your merge of develop got us a bit closer, but we had another, different fail... I've triggered another attempt, and now it's another different fail again. 

Can I ask you and the team to take a look please? Thank you!

**Comment by jennifer-shehane** [comment]
Yeah, I think something gets weird in CircleCI sometimes when the tests are triggered from certain outside contributors that I haven't pinned down.

**Comment by GCHQDeveloper911** [comment]
Whoop! We got a full set of passes @jennifer-shehane  :)

File: packages/network/lib/pki.ts
Comment (flotwig) [inline packages/network/lib/pki.ts:138]: Refactor usage of this class to not rely on a global singleton instance.

File: packages/network/lib/index.ts
Comment (flotwig) [inline packages/network/lib/index.ts:16]: this is already in the `network` package, so this can be exported as just `pki`, not `networkPki`

File: packages/network/test/unit/pki_spec.ts
Comment (flotwig) [inline packages/network/test/unit/pki_spec.ts:1]: there seems to have been some kind of error here - this is not a spec, this is a line-for-line copy of `packages/network/lib/pki.ts`, this should be replaced with unit tests

**Review by flotwig — CHANGES_REQUESTED** [review]
Good PR! The API design looks solid, it is similar to what was proposed in the issue comments but a little closer to Node's APIs.

In general, this PR looks good, but it needs *tests*, including unit tests and an end-to-end test of the feature.

It looks like there were supposed to be some tests in `pki_spec.ts`, but they were replaced with the contents of `pki.ts` by mistake?

For an end-to-end test, you can create a dummy project and web server to test PKI with in `packages/server/test/e2e`. See the files there for examples of how to use the e2e testing utility to launch a Cypress fixture project and how to launch a web server, which you can then enable PKI on.

Please feel free to let me know if you have any questions on these points. It's exciting to see PKI support for Cypress becoming real :)

File: packages/network/test/unit/pki_spec.ts
Comment (GCHQDeveloper911) [reply packages/network/test/unit/pki_spec.ts:1]: I'm told the tests do exist, were omitted by mistake, and they'll be submitted soon!

File: packages/network/lib/index.ts
Comment (GCHQDeveloper911) [reply packages/network/lib/index.ts:16]: Hi @flotwig - we should be able to provide an update here early next week

File: packages/network/lib/pki.ts
Comment (GCHQDeveloper911) [reply packages/network/lib/pki.ts:138]: Hi @flotwig - we should be able to provide an update here early next week

**Comment by Nastaran200** [comment]
Will this feature be added as a free feature of cypress? and when is the expected release?

**Comment by flotwig** [comment]
> Will this feature be added as a free feature of cypress? and when is the expected release?

All features of the Test Runner are free.

Since this is a community PR and not being created by Cypress team, we can't offer an estimate on when this will be released.

**Comment by vodlogic** [comment]
Any update on progress for this PR @GCHQDeveloper911 ?  Until PKI support is implemented in Cypress we will sadly need to stick with Puppeteer

**Comment by GCHQDeveloper911** [comment]
Unfortunately other commitments have taken priority in recent days. I hope to have an update soon!

**Comment by vodlogic** [comment]
> Unfortunately other commitments have taken priority in recent days. I hope to have an update soon!

Understood. In the meantime are you able to share details on how I can test your branch of code in our own project? I've built it but I haven't managed to get it to work with the app in the cache path as yet.

Thanks!

**Comment by GCHQDeveloper911** [comment]
While this work is ongoing, and yet to become part of Cypress officially, here's some steps for folks that might want to try building the modified 6.6.0 source as currently on this fork, and running cypress that way. Bear in mind this is development code still, it should work fine, but YMMV, caveat emptor and all that!

Note: These instructions assume you have cloned the appropriate cypress source repo, and have yarn installed

1. Build
Install dependencies:
    ```
    yarn
    ```
    
Now build the binaries for the required target OS - Note!! Requires that the build agent has the same OS! When prompted, choose to just override the currently deployed version.
```
yarn binary-build --platform=linux
```
or
```
yarn binary-build --platform=win32
```
Finally, produce a cypress.zip output:
```
yarn binary-zip
```
Installing
Update your test application to Cypress v6.6.0, this will install the Cypress binaries and the related JavaScript interface(s)
```
npm install --save-dev cypress@6.6.0
```
Remove the 6.6.0 binaries from the follow path as appropriate. Optionally, you can also specify the child 6.6.0 folder as appropriate to only remove the binaries for that version:
Linux: ~/.cache/Cypress
Windows: /AppData/Local/Cypress/Cache

Now install the newly created binaries (bash syntax shown - adapt as appropriate for your platform). Note that you'll need to copy the cypress.zip created above to the current folder for this step (or modify to specify an absolute path, not relative).
```
CYPRESS_INSTALL_BINARY=$(pwd)/cypress.zip npm install --save-dev cypress@6.6.0
```

**Comment by GCHQDeveloper911** [comment]
@flotwig - Hopefully your comments above have all been addressed, bar the end to end testing - that'll come soon... ish!

File: packages/network/lib/pki.ts
Comment (GCHQDeveloper911) [reply packages/network/lib/pki.ts]: @flotwig Are you happy this is resolved in the latest commit? Can this conversation be resolved?

File: packages/network/lib/index.ts
Comment (GCHQDeveloper911) [reply packages/network/lib/index.ts]: @flotwig Are you happy this is resolved in the latest commit? Can this conversation be resolved?

File: packages/network/test/unit/pki_spec.ts
Comment (GCHQDeveloper911) [reply packages/network/test/unit/pki_spec.ts]: @flotwig Are you happy this is resolved in the latest commit? Can this conversation be resolved?

**Comment by GCHQDeveloper911** [comment]
@jennifer-shehane Can you help here - I pushed a further commit 5 days ago, but the CI checks seem to have just hung?

**Comment by steverb1** [comment]
Any estimate on when this will be released?

**Comment by GCHQDeveloper911** [comment]
@steverb1 We're waiting for a response on the CI issues blocking the latest commit, and hopefully @flotwig will confirm the issues previously raised are now resolved - and then maybe we'll see this merged in soon :)

**Comment by flotwig** [comment]
@GCHQDeveloper911 can you please resolve the merge conflicts in this PR against `develop`? Looks like the big 7.0 release merge introduced some conflicts.

**Comment by GCHQDeveloper911** [comment]
This is a bit broken at the moment - will look to get things up and running, in line with 7.1.0 as soon as possible.

**Comment by GCHQDeveloper911** [comment]
@flotwig / @jennifer-shehane Can you advise here please - circleCI and appveyor builds are failing, but the errors appears to be nothing to do with the PKI changes we've made - @cypress/design-system?

**Comment by flotwig** [comment]
> @flotwig / @jennifer-shehane Can you advise here please - circleCI and appveyor builds are failing, but the errors appears to be nothing to do with the PKI changes we've made - @cypress/design-system?

The merge appears to have been done incorrectly, the diff in GitHub is showing a bunch of files added that should not have been.

**Comment by GCHQDeveloper911** [comment]
Hi @flotwig - can you take a look here again please. These failures again seem to be nothing to do with the PKI work. Indeed we have a chrome failure above that passed on FF, an FF failure that passed on chrome, a couple of test timeouts (10 min inactive), and an actual test failure that seems to have nothing to do with the PKI work. Also, on a previous CI run earlier today only one of these fails happened - and nothing major has changed since that run?

**Comment by GCHQDeveloper911** [comment]
Could someone with the appropriate permissions please re-run the above failed checks? As far as I can tell they should pass, but there is a degree of unpredictability in those tests. I don't believe the failures are linked to the work done in the PR.

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 Yes, it may be flake. I reran the failing tests. We're going to try to do a final technical review soon and also have our team look into documentation, API finalization for the user experience.

**Comment by GCHQDeveloper911** [comment]
Thanks @jennifer-shehane, great news!

**Comment by GCHQDeveloper911** [comment]
@flotwig - Are you happy for your comments from 3rd March to be marked as resolved?

**Comment by jennifer-shehane** [comment]
Hey @GCHQDeveloper911, in an effort to try to save time in going back and forth, our team is waiting for the PR to be in a state that is complete on your side, with all of the code done and the CI checks passing before doing a re-review. Let us know when it's in a state for a full re-review. Thanks!

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane We promise to leave it alone now! A few CI fails currently, but I believe they are all due to flake? Cheers

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 The `_deconstructPacket` error in the desktop-gui is flake. I'm not sure about the other 2 specs. I'll rerun, but I don't think they typically fail.

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:289]: It would be nice to supply more details here.
```suggestion
                "description": "Requests for URLs matching this minimatch pattern will use the supplied client certificate.",
```

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:293]: Generally we want to discourage users to hard-code absolute paths in their projects, as it makes it harder for users to share tests with other developers.

```suggestion
                "description": "Path(s) to CA file(s) to validate certs against, relative to project root.",
```

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:308]: Same absolute path feedback

```suggestion
                          "description": "Path to the certificate, relative to project root",
```

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:312]:
```suggestion
                          "description": "Path to the private key, relative to project root",
```

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:318]: Hmm, why is it preferred to have a passphrase in a file vs. including it in this config?

File: cli/schema/cypress.schema.json
Comment (flotwig) [inline cli/schema/cypress.schema.json:330]: same absolute path feedback

```suggestion
                          "description": "Path to the .pfx file, relative to project root",
```

File: packages/network/lib/index.ts
Comment (flotwig) [inline packages/network/lib/index.ts:7]: i know we already named this from `networkPki` to `pki`, but in line with the earlier comment, this code module should probably be called `clientCertificates`, not `pki`

File: packages/network/lib/pki.ts
Comment (flotwig) [inline packages/network/lib/pki.ts:134]: everything named `SomethingPkiSomething` needs the pki removed... so this is now `ClientCertificateStore`

File: packages/network/lib/pki.ts
Comment (flotwig) [inline packages/network/lib/pki.ts:139]:
```suggestion
      'ClientPkiCertificateStore::addPkiCertificatesForUrl: "%s"',
```

File: packages/network/lib/pki.ts
Comment (flotwig) [inline packages/network/lib/pki.ts:165]:
```suggestion
    const parsedUrl = new ParsedUrl(requestUrl.href)
    const matchingCerts = this._urlPkicertificates.filter((cert) => {
      return PkiUrlMatcher.matchUrl(parsedUrl, cert.matchRule)
    })

```

File: packages/network/test/unit/agent_spec.ts
Comment (flotwig) [inline packages/network/test/unit/agent_spec.ts:170]: hmm, we don't necessarily want to run all of our HTTPS tests with a client certificate, this should be moved to its own test/own set of tests

File: packages/network/test/unit/agent_spec.ts
Comment (flotwig) [inline packages/network/test/unit/agent_spec.ts:215]: ditto with earlier, let's make a new test for this instead of modifying the HTTPS tests that exist

File: packages/server/lib/pki.ts
Comment (flotwig) [inline packages/server/lib/pki.ts:177]: this should not be necessary, cypress server should be running with pwd equal to the project root already, can you confirm this is unneeded and remove it?

File: packages/server/lib/pki.ts
Comment (flotwig) [inline packages/server/lib/pki.ts:234]: this logic, and other logic in this file that uses `node-forge`, seems like it belongs in `@packages/network`, not here. would you agree?

**Review by flotwig — CHANGES_REQUESTED** [review]
Looking good so far! I made some comments about things to change, and then we can probably button this up with 1 more round of review.

**Comment by GCHQDeveloper911** [comment]
Thanks @flotwig / @jennifer-shehane - Will get those suggestions and CI fixes sorted next week. Cheers

**Comment by GCHQDeveloper911** [comment]
Hi @flotwig , some, not all of your changes addressed now in theory. Can you confirm please?

**Comment by flotwig** [comment]
> Hi @flotwig , some, not all of your changes addressed now in theory. Can you confirm please?

@GCHQDeveloper911 To make the best use of review time for both of us, it's best for me to only review when all changes have been definitely addressed, so let me know when you are ready for that. I do see your new commits :)

**Comment by nthali** [comment]
Hello, curious to see how this is progressing, and if there's any way we can help.
we are really waiting for this feature to become available as soon as possible

**Comment by danm** [comment]
Same. @GCHQDeveloper911 happy to pick something up here too. Happy to jump on a call to see where you got to if that is easier and quicker.

**Comment by GCHQDeveloper911** [comment]
Hi all, 

we're working on getting review issues resolved, and a (hopeful!) final push up to this PR this week. Thanks for your patience and interest!

**Comment by GCHQDeveloper911** [comment]
Hi @flotwig - CI seems a bit hosed generally at the moment, however I think all the comments you raised have now been addressed. Can you confirm please?

Thanks!

**Comment by flotwig** [comment]
> Hi @flotwig - CI seems a bit hosed generally at the moment, however I think all the comments you raised have now been addressed. Can you confirm please?

Yeah, it seems like this branch has changed the permissions on a bunch of files, which broke the build:

image

@GCHQDeveloper911 I don't have permissions to push to your repository, but I have pushed up a commit that you can cherry-pick that revert the permissions changes:

https://github.com/cypress-io/cypress/commit/fb87d7f6ad1a4284fab5f4944ff58ab71e1438d1

If you check this box, I can push it directly to your branch: https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

**Comment by GCHQDeveloper911** [comment]
@flotwig - That should be better?

**Review by flotwig — DISMISSED** [review]
@GCHQDeveloper911 I made some commits for minor cleanup things + fixing CI, and this looks good.

This feature will require documentation before being released. Could you open a companion PR in https://github.com/cypress-io/cypress-documentation describing the `clientCertificates` API? https://docs.cypress.io/guides/references/configuration.html would probably be the most appropriate place.

**Comment by GCHQDeveloper911** [comment]
Just to post an update - we're working on the docs. Will link to the appropriate PR here in due course!

**Comment by cowboy** [comment]
This looks great! My only concern is that in the schema changes, I see the phrase **"relative to project root"** and I want to make sure it's totally clear what that means to an end-user.

Do we have a clear definition for what the "project root" is? (And is it defined by a specific config setting, or is it the location of `cypress.json`, or something else?)

**Comment by flotwig** [comment]
> Do we have a clear definition for what the "project root" is? (And is it defined by a specific config setting, or is it the location of `cypress.json`, or something else?)

There are other usages, for example, `projectRoot` is available on the plugin `config` object: https://docs.cypress.io/api/plugins/writing-a-plugin#config and it means the same thing it's meant to say in this PR.

**Comment by cowboy** [comment]
> There are other usages, for example, `projectRoot` is available on the plugin `config` object: https://docs.cypress.io/api/plugins/writing-a-plugin#config and it means the same thing it's meant to say in this PR.

Instead of "relative to project root" can the schema descriptions be updated to say "relative to projectRoot" ?

**Comment by jennifer-shehane** [comment]
@cowboy I usually refer to the project root as the directory that the `cypress.json` is located, but then again - people can set up a different file/location for the configFile. But this is a pretty common term in the Cypress project and our error messages if you search for `project root' in the `cypress` directly.

**Comment by cowboy** [comment]
> @cowboy I usually refer to the project root as the directory that the `cypress.json` is located, but then again - people can set up a different file/location for the configFile. But this is a pretty common term in the Cypress project and our error messages if you search for `project root` in the `cypress` repo.

Ok!

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 Looks like there's some linting issues with the latest commit.

**Comment by jennifer-shehane** [comment]
Hey @GCHQDeveloper911 Are you able to spend some time to fix the conflicts and failing linting/specs? This is the final step before we are ready to merge and release it.

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane Sorry for the delay - we should be able to get this sorted today

**Comment by GCHQDeveloper911** [comment]
Hi @jennifer-shehane - we believe it's all good now. I suspect the 2 CI fails are just flake? Many thanks for your ongoing patience and support here!

**Comment by jennifer-shehane** [comment]
@GCHQDeveloper911 Great, thanks! We'll give it a last look over soon.

**Review by flotwig — APPROVED** [review]
Looks good. Thank you for all the hard work @GCHQDeveloper911.

The remaining failure seems to be due to https://github.com/cypress-io/cypress/pull/17289, so we can proceed with merging.

**Comment by GCHQDeveloper911** [comment]
That's awesome! Thanks both of you! Looking fwd to Cypress 8! :)

**Comment by kchojhu** [comment]
I'm pretty sure my post here would have little value but I definitely appreciate if this can be release very soon.

**Comment by tcompactor** [comment]
Been looking for this feature for some time - big Thank You  for everyone involved with the implementation! esp: GCHQDeveloper911

**Comment by jennifer-shehane** [comment]
Released in [`8.0.0`](https://on.cypress.io/changelog#8-0-0).

This comment thread has been locked. If you are still experiencing an issue with this feature after upgrading to
Cypress v8.0.0, please [open a new issue](https://github.com/cypress-io/cypress/issues/new/choose).

**Pull request merged**
